pkg/apis/core/types.go

```diff
@@ -3510,17 +3510,39 @@ type LoadBalancerIngress struct {
 	Hostname string
 }
 
+const (
+	// MaxServiceTopologyKeys is the largest number of topology keys allowed on a service
+	MaxServiceTopologyKeys = 16
+)
+
 // IPFamily represents the IP Family (IPv4 or IPv6). This type is used
-// to express the family of an IP expressed by a type (i.e. service.Spec.IPFamily)
+// to express the family of an IP expressed by a type (i.e. Service.Spec.IPFamilies).
 type IPFamily string
 
+// IPFamilyPolicyType represents the dualstack-ness requested or required by a Service
+type IPFamilyPolicyType string
+
 const (
 	// IPv4Protocol indicates that this IP is IPv4 protocol
 	IPv4Protocol IPFamily = "IPv4"
 	// IPv6Protocol indicates that this IP is IPv6 protocol
 	IPv6Protocol IPFamily = "IPv6"
-	// MaxServiceTopologyKeys is the largest number of topology keys allowed on a service
-	MaxServiceTopologyKeys = 16
+	// SingleStack indicates that this service is required to have a single IPFamily
+	// The IPFamily assigned is based on the default IPFamily used by the cluster
+	// or as identified by Spec.IPFamilies field
+	SingleStack IPFamilyPolicyType = "SingleStack"
+	// PreferDualStack indicates that this service prefers dualstack (two IPFamilies) whenever
+	// the cluster is configured for dual stack. If the cluster is not configured
```

Comment:
What happens if the cluster IP Family is not match service.spec.ipfamilies? that is an error?

Reply:
depends. yes for most cases except headless w/o selector.